File: matlab/example_simple.m
clear all; close all; clc;

N = 4;
A = [1.01, 0, 0, 0; 0, 1, 0, 0; 0, 0, 0, -1/3; 0, 0, 1/3, 0];
U1 = zeros(N,1);
U1(1) = 1;

iter = 200;

beta = 1/4;
beta2 = 4/27;

spectral_gap = 0.01;

disp('A=');
disp(A);

xinit = rand(N,1) + 1i*rand(N,1);
disp('xinit=');
disp(xinit);

[x1, errs1] = powermethod(A, xinit, iter, U1);
disp('power method err=');
disp(errs1(end));

[x2, errs2] = momentum(A, xinit, iter, beta, U1);
disp('momentum method err=');
disp(errs2(end));

[x3, errs3] = momentum2(A, xinit, iter, beta2, U1);
disp('order 2 momentum method err=');
disp(errs3(end));

[x4, errs4] = momentum_dynamic(A, xinit, iter, U1);
disp('order 2 dynamic momentum method err=');
disp(errs4(end));

iters = 0:iter;

figure;
semilogy(iters, errs1,'DisplayName','power method');
hold on;
semilogy(iters, errs2,'DisplayName','momentum');
semilogy(iters, errs3,'DisplayName','order 2 momentum');
semilogy(iters, errs4,'DisplayName','order 2 dyn momentum');

asympt = exp(-iters*sqrt(spectral_gap));
semilogy(iters, 1.2*asympt*errs3(101)/asympt(101), '--', 'DisplayName','theory order 2 momentum');

ylim([1e-10, 10]);

legend;
xlabel('n');
ylabel('relative error');
hold off;

File: python/example_simple.py
import numpy as np
from powermethods import powermethod, momentum, momentum2, momentum_dynamic
import matplotlib.pyplot as plt

# simple 4x4 example with eigenvalues that lie within a deltoid
N = 4
A = np.array([[1.01, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, -1/3], [0, 0, 1/3, 0]])
U1 = np.zeros(N)
U1[0] = 1

# iteration count
iter = 200

# relative gap between magnitude of 1st and 2nd eigenvalues
# used to determine theoretical rate of convergence
spectral_gap = 0.01

print("A=\n",A)

xinit = (np.random.rand(N)+1j*np.random.rand(N)).reshape(-1,1)
print("xinit=\n",xinit)

# momentum parameters
beta = 1/4
beta2 = 4/27

# apply different power methods for comparison purposes
x1, errs1 = powermethod(A, xinit, iter, xtrue=U1)
print("power method err=\n",errs1[-1])

x2, errs2 = momentum(A, xinit, iter, beta, xtrue=U1)
print("momentum method err=\n",errs2[-1])

x3, errs3 = momentum2(A, xinit, iter, beta2, xtrue=U1)
print("order 2 momentum method err=\n",errs3[-1])

x4, errs4 = momentum_dynamic(A, xinit, iter, xtrue=U1)
print("order 2 dynamic momentum method err=\n",errs4[-1])

# plot results
iters = np.arange(iter+1)
plt.subplots()
plt.semilogy(iters, errs1, '-', marker='x', markevery=iter//10, label = 'power method')
plt.semilogy(iters, errs2, '-', marker='s', markevery=iter//10, label = f'momentum ($\\beta = {beta:.3f}$)')
plt.semilogy(iters, errs3, '-', marker='o', markevery=iter//10, label = 'order 2 momentum ($\\beta = 4/27$)')
plt.semilogy(iters, errs4, '-', marker='*', markevery=iter//10, label = 'order 2 dyn momentum')

# plot theoretical asymptotic convergence as well
asympt = np.pow(1+np.sqrt(spectral_gap), -iters)
plt.semilogy(iters, asympt * errs3[-1]/asympt[-1], '--', label = r"$O((1+\sqrt{\varepsilon})^{-N})$")
plt.ylim(1e-10, 10)

plt.legend()
plt.xlabel("n")
plt.ylabel("relative error")
plt.tight_layout()
#plt.savefig("relerr_simple.eps")
plt.show()
